File: py/todoosy/formatter.py
"""
Todoosy Formatter
"""

from .parser import parse
from .types import ItemNode, ItemMetadata


def format_metadata(metadata: ItemMetadata) -> str:
    """Format metadata as a canonical string."""
    parts: list[str] = []

    if metadata.due:
        parts.append(f"due {metadata.due}")

    if metadata.priority is not None:
        parts.append(f"p{metadata.priority}")

    if metadata.estimate_minutes is not None:
        minutes = metadata.estimate_minutes
        if minutes % 480 == 0 and minutes >= 480:
            parts.append(f"{minutes // 480}d")
        elif minutes % 60 == 0 and minutes >= 60:
            parts.append(f"{minutes // 60}h")
        else:
            parts.append(f"{minutes}m")

    return f"({' '.join(parts)})" if parts else ''


def format_item_line(item: ItemNode, indent: int = 0) -> str:
    """Format a single item line."""
    indent_str = '  ' * indent
    meta_str = format_metadata(item.metadata)
    title_with_meta = f"{item.title_text} {meta_str}" if meta_str else item.title_text

    if item.type == 'heading':
        hashes = '#' * (item.level or 1)
        return f"{hashes} {title_with_meta}"

    return f"{indent_str}- {title_with_meta}"


def format_comments(comments: list[str], is_list_item: bool, indent: int) -> list[str]:
    """Format comments with proper indentation."""
    if not comments:
        return []

    if is_list_item:
        indent_str = '  ' * (indent + 1)
        return [f"{indent_str}{c}" for c in comments]

    return list(comments)


def format(text: str) -> str:
    """Format a todoosy document."""
    result = parse(text)
    ast = result.ast
    lines: list[str] = []
    item_map = {item.id: item for item in ast.items}

    # Track Misc section
    misc_section_id = None
    for item in ast.items:
        if item.type == 'heading' and item.title_text == 'Misc' and item.level == 1:
            misc_section_id = item.id
            break

    def format_item(item_id: str, list_indent: int = 0, is_under_misc: bool = False) -> None:
        item = item_map[item_id]

        # Skip Misc section during normal iteration
        if item.id == misc_section_id and not is_under_misc:
            return

        # Add blank line before headings (except at start)
        if item.type == 'heading' and lines and lines[-1] != '':
            lines.append('')

        lines.append(format_item_line(item, list_indent))

        # Add blank line after heading before comments or children
        if item.type == 'heading':
            lines.append('')

        # Add comments
        formatted_comments = format_comments(
            item.comments,
            item.type == 'list',
            list_indent
        )
        lines.extend(formatted_comments)

        # Add blank line after heading comments before children
        if item.type == 'heading' and item.comments and item.children:
            lines.append('')

        # Format children
        for child_id in item.children:
            child = item_map[child_id]
            if child.type == 'list':
                next_indent = 0 if item.type == 'heading' else list_indent + 1
                format_item(child_id, next_indent, is_under_misc)
            else:
                format_item(child_id, 0, is_under_misc)

    # Format all root items except Misc
    for root_id in ast.root_ids:
        if root_id != misc_section_id:
            format_item(root_id, 0, False)

    # Add Misc section at the end
    if lines and lines[-1] != '':
        lines.append('')
    lines.append('# Misc')

    # Add Misc items if they exist
    if misc_section_id:
        misc_item = item_map[misc_section_id]
        if misc_item.comments:
            lines.append('')
            lines.extend(misc_item.comments)
        if misc_item.children:
            lines.append('')
            for child_id in misc_item.children:
                child = item_map[child_id]
                lines.append(format_item_line(child, 0))
                formatted_comments = format_comments(
                    child.comments,
                    child.type == 'list',
                    0
                )
                lines.extend(formatted_comments)

    return '\n'.join(lines) + '\n'

File: js/src/formatter.ts
/**
 * Todoosy Formatter
 */

import { parse } from './parser.js';
import type { ItemNode, ItemMetadata, Scheme } from './types.js';

function parseMiscLocation(misc: string): { filename: string; heading: string } {
  const slashIndex = misc.indexOf('/');
  if (slashIndex === -1) {
    return { filename: misc, heading: 'Misc' };
  }
  return {
    filename: misc.substring(0, slashIndex),
    heading: misc.substring(slashIndex + 1),
  };
}

function formatMetadata(metadata: ItemMetadata): string {
  const parts: string[] = [];

  if (metadata.due) {
    const softPrefix = metadata.due_soft ? '~' : '';
    parts.push(`due ${softPrefix}${metadata.due}`);
  }

  if (metadata.progress) {
    parts.push(metadata.progress);
  }

  if (metadata.priority !== null) {
    parts.push(`p${metadata.priority}`);
  }

  if (metadata.estimate_minutes !== null) {
    const minutes = metadata.estimate_minutes;
    if (minutes % 480 === 0 && minutes >= 480) {
      parts.push(`${minutes / 480}d`);
    } else if (minutes % 60 === 0 && minutes >= 60) {
      parts.push(`${minutes / 60}h`);
    } else {
      parts.push(`${minutes}m`);
    }
  }

  return parts.length > 0 ? `(${parts.join(' ')})` : '';
}

function formatItemLine(item: ItemNode, indent: number = 0): string {
  const indentStr = '  '.repeat(indent);
  const metaStr = formatMetadata(item.metadata);
  const titleWithMeta = metaStr ? `${item.title_text} ${metaStr}` : item.title_text;

  if (item.type === 'heading') {
    const hashes = '#'.repeat(item.level || 1);
    return `${hashes} ${titleWithMeta}`;
  }

  return `${indentStr}- ${titleWithMeta}`;
}

function formatComments(comments: string[], isListItem: boolean, indent: number): string[] {
  if (comments.length === 0) return [];

  if (isListItem) {
    // Indent is the list item's indent level; comments need +1 level
    const indentStr = '  '.repeat(indent + 1);
    return comments.map(c => `${indentStr}${c}`);
  }

  // Heading comments are not indented
  return comments;
}

export function format(text: string, scheme?: Scheme, filename?: string): string {
  const { ast } = parse(text);
  const lines: string[] = [];
  const itemMap = new Map<string, ItemNode>();

  for (const item of ast.items) {
    itemMap.set(item.id, item);
  }

  // Determine misc location from scheme or use default
  const miscLocation = parseMiscLocation(scheme?.misc ?? 'todoosy.md/Misc');
  // If no filename provided, assume it could be the misc file (backward compatibility)
  const isMiscFile = filename === undefined || filename === miscLocation.filename;

  // Determine formatting style: roomy (default), balanced, or tight
  const formattingStyle = scheme?.formatting_style ?? 'roomy';

  // Track Misc section
  let miscSectionId: string | null = null;

  // Find existing Misc section (using configured heading name)
  for (const item of ast.items) {
    if (item.type === 'heading' && item.title_text === miscLocation.heading && item.level === 1) {
      miscSectionId = item.id;
      break;
    }
  }

  // Helper to determine if we should add blank line before a heading
  function shouldAddBlankBefore(item: ItemNode): boolean {
    if (item.type !== 'heading') return false;
    if (formattingStyle === 'tight') return false;
    if (formattingStyle === 'balanced') return item.level === 1;
    return true; // roomy
  }

  // Helper to determine if we should add blank line after a heading
  function shouldAddBlankAfter(item: ItemNode): boolean {
    if (item.type !== 'heading') return false;
    if (formattingStyle === 'tight') return false;
    if (formattingStyle === 'balanced') return item.level === 1;
    return true; // roomy
  }

  // Helper to format an item and its subtree
  function formatItem(id: string, listIndent: number = 0, isUnderMisc: boolean = false): void {
    const item = itemMap.get(id)!;

    // Skip Misc section during normal iteration (we'll add it at the end)
    if (item.id === miscSectionId && !isUnderMisc) {
      return;
    }

    // Add blank line before headings (except at start), based on style
    if (shouldAddBlankBefore(item) && lines.length > 0) {
      if (lines[lines.length - 1] !== '') {
        lines.push('');
      }
    }

    lines.push(formatItemLine(item, listIndent));

    // Add blank line after heading before comments or children, based on style
    if (shouldAddBlankAfter(item)) {
      lines.push('');
    }

    // Add comments
    const formattedComments = formatComments(
      item.comments,
      item.type === 'list',
      listIndent
    );
    lines.push(...formattedComments);

    // Add blank line after heading comments before children
    if (item.type === 'heading' && item.comments.length > 0 && item.children.length > 0) {
      if (shouldAddBlankAfter(item)) {
        lines.push('');
      }
    }

    // Format children
    for (const childId of item.children) {
      const child = itemMap.get(childId)!;
      if (child.type === 'list') {
        // List items under a heading start at indent 0
        // Nested list items increment indent
        const nextIndent = item.type === 'heading' ? 0 : listIndent + 1;
        formatItem(childId, nextIndent, isUnderMisc);
      } else {
        formatItem(childId, 0, isUnderMisc);
      }
    }
  }

  // Format all root items except Misc
  for (const rootId of ast.root_ids) {
    if (rootId !== miscSectionId) {
      formatItem(rootId, 0, false);
    }
  }

  // Add Misc section at the end (only for the misc file)
  if (isMiscFile) {
    // Add blank line before Misc heading based on style
    const miscItem = miscSectionId ? itemMap.get(miscSectionId) : null;
    const shouldAddBlankBeforeMisc = formattingStyle !== 'tight' && (formattingStyle === 'roomy' || formattingStyle === 'balanced');
    if (lines.length > 0 && lines[lines.length - 1] !== '' && shouldAddBlankBeforeMisc) {
      lines.push('');
    }
    lines.push(`# ${miscLocation.heading}`);

    // Add blank line after Misc heading based on style
    const shouldAddBlankAfterMisc = formattingStyle !== 'tight' && (formattingStyle === 'roomy' || formattingStyle === 'balanced');

    // Add Misc items if they exist
    if (miscSectionId) {
      const miscItemNode = itemMap.get(miscSectionId)!;
      if (miscItemNode.comments.length > 0) {
        if (shouldAddBlankAfterMisc) {
          lines.push('');
        }
        lines.push(...miscItemNode.comments);
      }
      if (miscItemNode.children.length > 0) {
        if (shouldAddBlankAfterMisc) {
          lines.push('');
        }
        for (const childId of miscItemNode.children) {
          // Format misc children - they start at indent 0
          const child = itemMap.get(childId)!;
          lines.push(formatItemLine(child, 0));
          const formattedComments = formatComments(
            child.comments,
            child.type === 'list',
            0
          );
          lines.push(...formattedComments);
        }
      }
    }
  }

  return lines.join('\n') + '\n';
}
